### ISSUE TITLE
Newline character in vector tiles is ignored in text label

### ISSUE THREAD

**Issue by nextstopsun** [mapbox/mapbox-gl-js#6064]
This issue comes from this one: https://github.com/mapbox/mapbox-gl-js/pull/3179

I've got a vector tile server with features having multiline text properties. I'm looking for a way to tell MGL to break lines at special characters. I've tried using both `\n` and `char(13)` for line break, but neither works. `char(13)` is ignored, while `\n` gets displayed as is.

I'd be great to have an option like `text-newline-character`, so one can set a specific character to break lines at.

![](https://user-images.githubusercontent.com/966321/35384521-9dad6b14-01d6-11e8-9082-1b195c0dd263.png)

**Comment by ryanhamley** [comment]
If you have the ability to use newer JS features, a template literal multi-line string will be rendered as expected in a text label.

**Comment by nextstopsun** [comment]
@ryanhamley Can you please give more explanation?

**Comment by KempWatson** [comment]
I'm having this issue too - in Studio, should I not be able to enter something like "\n" in a text field, and get a newline? Newline _is_ a character... I don't want JS, this should be directly enterable in Studio text fields.

**Comment by james-wagner11** [comment]
Not sure if it applies to your cases, but [this answer helped me](https://stackoverflow.com/questions/42786557/how-can-i-inject-custom-line-breaks-on-a-symbol-layer-label/42798431#42798431).

According to that, setting text-max-width to 0 disables all wrapping (including \n). Setting it to Infinity fixed newline wrapping in my case.

**Comment by ryanhamley** [comment]
Hi @nextstopsun are you still experiencing this issue? If so, are you able to provide a minimal setup in JSBin that reproduces the error? I tried investigating and found that `'hello \n world'` renders with the new line. Even using Cyrillic characters rendered as expected `'Лорем\nипсум\nдолор\nсит\nамет'`. I'm using Chrome 66.0.3359.181 on a Mac. If you can share your browser and OS info with a working demo of the bug that would be very helpful in solving this issue.

**Comment by jfirebaugh** [comment]
@nextstopsun You need to use the newline character (hexadecimal `0x0a`, `char(10)`), rather than the carriage return character (hexadecimal `0x0d`, `char(13)`).

**Comment by KempWatson** [comment]
@jfirebaugh Can I ask why this is closed? None of this works for me.

**Comment by kdanaher** [comment]
Chiming in here as I am experiencing the same thing. Newline character is baked directly into my vector tiles and doesn't seem to trigger a newline on the tileset when pulled into Studio.

**Comment by NickPepper** [comment]
The issue still exists. Chrome ver 61.0.3163.100 (64-bit), MacOS ver 10.14.5
No way to use linebreaks.

Can I also ask why this issue is closed just like @jfirebaugh did?

**Comment by ryanhamley** [comment]
Hi @KempWatson @kdanaher @NickPepper I cannot reproduce the issues that you are reporting. Here's [an example](https://jsbin.com/gulodivabo/edit?html,output) of a newline character being rendered properly from data embedded in a vector tile. This was created using Mapbox Studio. This is the GeoJSON I used to create the vector tile.

```
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {
        "text": "hello\nworld"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          -77.03651905059814,
          38.89332088074671
        ]
      }
    }
  ]
}
```

[This example](https://jsbin.com/cuyuroj/edit?html,output) demonstrates using a newline character in a GeoJSON source rather than text that's embedded into a vector tile. Both the `\n` character and `String.fromCharCode(10)` create newlines as expected. `\r` and `String.fromCharCode(13)` will not because that's a carriage return rather than a newline.

If anyone believes there's a bug in how newlines are being handled, please include a demonstration of the issue so that we can determine if there's something unexpected happening. If you need general support on how to get newlines working in your style, you can contact [Mapbox Support](https://support.mapbox.com/hc/en-us). Thanks everyone!

**Comment by ryanhamley** [comment]
I've got an update that might help some folks who find this issue. If you're parsing JSON, you will need to escape the entire newline character before parsing in order for the newline to be rendered correctly. So it would need to be `\\n` before running `JSON.parse`. See [this Stack Overflow question](https://stackoverflow.com/questions/11591784/parsing-json-containing-new-line-characters).

**Comment by KempWatson** [comment]
Ryan, sadly that's not relevant to this issue... it's not about parsing
JSON, it's about entering a string with a carriage return into Mapbox
Studio, and generating tiles from that. \n, \r, \n\r, \r\n, \\n, nothing
works. Should be easy, but this has been a problem and an open issue for
quite a long time. I have tested again last night, the problem persists.

JSBin will not help, the problem is not in coding, it is in Mapbox itself.

K. Watson

**Comment by ryanhamley** [comment]
@KempWatson See https://github.com/mapbox/mapbox-gl-js/issues/8575 which was reported yesterday. There is a bug where if `text-max-width` is set to zero, newlines are ignored. Otherwise, newline characters do seem to work as expected as I demonstrated in my comment above.

Note though that the carriage return does not render a newline. My testing shows that `\n`, `\r\n`, and `\n\r` produce a line break. `\r` does not because GL JS considers the carriage return to be whitespace, not a breakable character as can be seen [in this code](https://github.com/mapbox/mapbox-gl-js/blob/master/src/symbol/shaping.js#L219). The library [forces a line break on the newline character](https://github.com/mapbox/mapbox-gl-js/blob/master/src/symbol/shaping.js#L281-L284) (`\n` or `String.fromCharCode(10)`). It would be possible to also force a break on carriage returns (`\r\` or `String.fromCharCode(13)`) but this would require some discussion.

For one thing, making that change would probably be a breaking change which would have the potential to alter how some users' maps are rendered and it would likely be very hard for those users to figure out why they suddenly had new line breaks. But it's also not obvious that a carriage return *should* produce a line break. Historically, a carriage return moved a mechanical carriage back to the start of the existing line and it had to be combined with a line feed to move to the next line (hence the `\r\n` construct in that particular order). I'm not an expert here but my understanding is that Windows has generally used `\r\n` while Unix and OSX both use `\n` so causing a breaking change in the library to force a break on just `\r` seems unnecessary to me. If anyone has a minimal example that shows a newline character not being rendered when the `text-max-width` is set to > 0, let us know as that would be helpful in debugging any further issues with this.

**Comment by KempWatson** [comment]
I've attached some screenshots here, of both the dataset and the style
settings.

I've tried everything, nothing makes a line break. Note that the "Max
width" is set to 18em, but the fix above still does not work.

Kemp

**Comment by ghigio2000** [comment]
If I use the \n into the "text-field" definition it works. For example:
["concat",["get","name"],"\n(",["get","ele"],")"]

if I have the "name" field with a \n inside then it does not work. For example:
"New \nYork"
not working also
"New \\nYork"

New ideas?

Giorgio

**Comment by KempWatson** [comment]
To correct the above: this is about newlines, not technically carriage
returns. How does one define a multi-line string in a field in the portal?
\r, \n, \n\r, \r\n, <br>, html entities, all fail.

Kemp Watson

**Comment by KempWatson** [comment]
This is not the issue - nothing to do with JavaScript, or PostGIS, or
anything else. Just simply the operation of the Mapbox Studio console at
https://www.mapbox.com/studio.

Kemp Watson

**Comment by ryanhamley** [comment]
Hey @KempWatson it turns out that you are correct. There's a bug in Studio when entering newline characters directly into a text field. It seems like the text field is being sanitized. I've filed a bug with the Studio team https://github.com/mapbox/studio/issues/12212

In the meantime, there are two options to work around this. Once you enter the text in the text-field, you can click the `Edit JSON` button at the bottom of the column (see the image below). This will show you the output JSON which will have an extra `\` in it. Remove that slash and your newline should work as expected. For example, if you entered `Hello\nWorld` in the text field, the JSON will be `Hello\\nWorld`.

<img width="246" alt="Screen Shot 2019-10-10 at 12 29 08 PM" src="https://user-images.githubusercontent.com/4523080/66600174-c9b1cc00-eb59-11e9-9e33-5db4b4e93832.png">

You can also click on `Use a formula` and use the `&` to concat your strings with a newline character if you'd prefer to use an expression.

<img width="236" alt="Screen Shot 2019-10-10 at 12 31 04 PM" src="https://user-images.githubusercontent.com/4523080/66600239-f534b680-eb59-11e9-9fea-25e0dfc5cea0.png">

Thanks for reporting this and being persistent. Hopefully this will solve your issue.

**Comment by KempWatson** [comment]
Thanks Ryan... I'd swear I tried the first workaround to no avail, but not
the second. I'll try both tonight and report back here,
Kemp

**Comment by ramz211** [comment]
Hi @KempWatson @ryanhamley @ghigio2000 any success? This is still an issue in STUDIO. A /n in the text field box in Studio previews fine, and enters fine when viewing that json, but does not render with a line break, so frustrating...

**Comment by ryanhamley** [comment]
@ramz211 do you have an example you can point us to? either by creating a map or sharing the style you created

**Comment by KempWatson** [comment]
Ryan, it's really simple, just try to make a two-line text string in Mapbox
Studio. Can't do it... it's not about the map, it's about Studio.

Kemp Watson

**Comment by ramz211** [comment]
@ryanhamley here are screen shots of a \n previewing fine in Studio, but on Render it ignores the line breaks. In Studio, if you click on View Json, it looks fine, but when you go back to the Text Field boxes all \n have been removed

![Screenshot_2020-05-29 Studio](https://user-images.githubusercontent.com/64946725/83458469-77d6f280-a417-11ea-933b-b7ae9d9f7234.png)
![Screenshot_2020-05-29 Studio(1)](https://user-images.githubusercontent.com/64946725/83458482-7efe0080-a417-11ea-946f-8d43da984320.png)
![Screenshot_2020-05-29 Studio(2)](https://user-images.githubusercontent.com/64946725/83458505-86bda500-a417-11ea-9346-004c3af69e8c.png)
![Screenshot_2020-05-29 Monochrome2](https://user-images.githubusercontent.com/64946725/83458532-8c1aef80-a417-11ea-8d33-eaddded9346a.png)

**Comment by KempWatson** [comment]
> I'm seeing this issue as well. But only when the symbol-placement is set to line-center or line. If set to point newlines are displayed correctly.
> 
> http://jsfiddle.net/ghk3jqe2/3/

Again, this has NOTHING to do with Mapbox GL, it's STUDIO. 
4 years now and still not fixed, and bug issue 12212 which was supposed to address this bug no longer even exists. 
What gives, Mapbox?

**Comment by fraser-langton** [comment]
It wasn't immediately apparent but this was fixed in #12377 
release [v2.11.1](https://github.com/mapbox/mapbox-gl-js/releases/tag/v2.11.1)
